driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java

```diff
@@ -187,445 +144,603 @@ private void logError(InetSocketAddress address, Throwable exception) {
+                                    retriesByPolicy);
+
+                                if (metricsEnabled()) {
+                                    if (retry.getType() == Type.RETRY)
+                                        metrics().getErrorMetrics().getRetriesOnUnavailable().inc();
+                                    if (retry.getType() == Type.IGNORE)
+                                        metrics().getErrorMetrics().getIgnoresOnUnavailable().inc();
+                                }
+                                break;
+                            case OVERLOADED:
+                                release(connection);
+                                // Try another node
+                                logger.warn("Host {} is overloaded, trying next host.", connection.address);
+                                DriverException overloaded = new DriverException("Host overloaded");
+                                logError(connection.address, overloaded);
+                                if (metricsEnabled())
+                                    metrics().getErrorMetrics().getOthers().inc();
+                                retry(false, null);
+                                return;
+                            case SERVER_ERROR:
+                                release(connection);
+                                // Defunct connection and try another node
+                                logger.warn("{} replied with server error ({}), trying next host.", connection.address, err.message);
+                                DriverException exception = new DriverException("Host replied with server error: " + err.message);
+                                logError(connection.address, exception);
+                                connection.defunct(exception);
+                                if (metricsEnabled())
+                                    metrics().getErrorMetrics().getOthers().inc();
+                                retry(false, null);
+                                return;
+                            case IS_BOOTSTRAPPING:
+                                release(connection);
+                                // Try another node
+                                logger.error("Query sent to {} but it is bootstrapping. This shouldn't happen but trying next host.", connection.address);
+                                DriverException bootstrapping = new DriverException("Host is bootstrapping");
+                                logError(connection.address, bootstrapping);
+                                if (metricsEnabled())
+                                    metrics().getErrorMetrics().getOthers().inc();
+                                retry(false, null);
+                                return;
+                            case UNPREPARED:
+                                // Do not release connection yet, because we might reuse it to send the PREPARE message (see write() call below)
```

Comment:
So this is the main reason why we released later for [JAVA-633](https://datastax-oss.atlassian.net/browse/JAVA-633) right?  So the net of the change for JAVA-666 is to release in all cases of onSet except when you get an UNPREPARE error and the prepared statement will be reprepared.  Just making sure I'm understanding correctly.